### PR TITLE
Improve manifest cloning

### LINKED ISSUE #3180: The ssh.py module should never create files in the host system
The ``ssh.py`` module has a couple of methods that create local copies of files onto the host system in order to upload them to a Satellite system. An example of this would be the files required to clone a manifest. Instead of downloading them onto the filesystem, and since they're only temporary for the methods being invoked, we should instead use the ``StringIo`` module to create in memory file objects and use them to only then create artifacts onto the remote/final destination system. Example:

```diff
diff --git a/robottelo/manifests.py b/robottelo/manifests.py
index e56e05b..4c8f642 100644
--- a/robottelo/manifests.py
+++ b/robottelo/manifests.py
@@ -7,6 +7,7 @@ import json
 import os
 import requests
 import shutil
+import StringIO
 import tempfile
 import uuid
 import zipfile
@@ -37,10 +38,10 @@ def retrieve(url):

     """
     response = requests.get(url)
-    fd, tempname = tempfile.mkstemp()
-    with os.fdopen(fd, "wb") as temp_file:
-        temp_file.write(response.content)
-    return tempname
+    output = StringIO.StringIO()
+    output.write(response.content)
+
+    return output


 def download_signing_key():
@@ -156,13 +157,13 @@ def clone(key_path=None, old_path=None):
     :rtype: str

     """
-    tempdir = tempfile.mkdtemp()
+    tempdir = StringIO.StringIO()
     consumer_zip_file = os.path.join(tempdir, "consumer_export.zip")
     if key_path is None:
         key_path = download_signing_key()
     if old_path is None:
         old_path = download_manifest_template()
-    new_path = get_tempfile()
+    new_path = StringIO.StringIO()
     shutil.copy(old_path, new_path)
     with zipfile.ZipFile(new_path) as oldzip:
         oldzip.extractall(tempdir)
```

### PR DESCRIPTION
**Pull request by elyezer**
Avoid creating files on the filesystem, instead use in-memory file like
object.

Closes #3180

### CODE REVIEW

**Comment by oshtaier** [comment]
Are going to run few tests?

**Comment by elyezer** [comment]
@oshtaier yes! Wasn't able to do that before because needed to setup my own Satellite with latest compose.

**Comment by elyezer** [comment]
Have run RHAI tests:

```
$ py.test tests/foreman/rhai/test_rhai.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 3 items 

tests/foreman/rhai/test_rhai.py ...

============================ 3 passed in 462.58 seconds =============================
```

Will run more to make sure manifest uploading is working from API, CLI and UI.

**Comment by elyezer** [comment]
```
$ py.test tests/foreman/api/test_subscription.py                       
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 3 items 

tests/foreman/api/test_subscription.py ...

============================= 3 passed in 64.47 seconds =============================
```

**Comment by elyezer** [comment]
```
$ py.test tests/foreman/cli/test_subscription.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 6 items 

tests/foreman/cli/test_subscription.py sF....

================== 1 failed, 4 passed, 1 skipped in 150.68 seconds ==================
```

The failure is the same as get on latest automation run `tests.foreman.cli.test_subscription.SubscriptionTestCase.test_positive_enable_manifest_reposet`.

**Comment by elyezer** [comment]
```
$ py.test tests/foreman/ui/test_subscription.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 3 items 

tests/foreman/ui/test_subscription.py ...

============================ 3 passed in 177.17 seconds =============================
```

**Comment by elyezer** [comment]
PR is ready for review.

File: robottelo/manifests.py [hunk 1]
```diff
@@ -17,160 +12,103 @@
...
+    """Manifest clonning utility class."""
+    def __init__(self, template=None, signing_key=None):
+        self.template = template
+        self.signing_key = signing_key
+
+    def _download_manifest_info(self):
+        """Download and cache the manifest information."""
+        self.template = requests.get(settings.fake_manifest.url).content
+        self.signing_key = requests.get(settings.fake_manifest.key_url).content
+        self.signature = PKCS1_v1_5.new(RSA.importKey(self.signing_key))
```
Comment (omaciel) [inline robottelo/manifests.py]: I guess we have never supported or don't want to support DSA?

File: robottelo/manifests.py [hunk 2]
```diff
@@ -17,160 +12,103 @@
...
+
+        :return: A file-like object (``BytesIO`` on Python 3 and
+            ``StringIO`` on Python 2) with the contents of the cloned
+            manifest.
+        """
+        if self.signing_key is None or self.template is None:
+            self._download_manifest_info()
+
+        template_zip = zipfile.ZipFile(six.BytesIO(self.template))
+        consumer_export_zip = zipfile.ZipFile(
```
Comment (omaciel) [inline robottelo/manifests.py]: I'm guessing that we want to extract the file named `consumer_export.zip` which is found within `self.template`? If this is correct, mind documenting it with a comment?

File: robottelo/manifests.py [hunk 3]
```diff
@@ -17,160 +12,103 @@
...
+                        name,
+                        consumer_export_zip.read(name)
+                    )
+
+        # Generate a new manifest.zip file with the generated
+        # consumer_export.zip and new signature.
+        manifest = six.BytesIO()
+        with zipfile.ZipFile(
+                manifest, 'w', zipfile.ZIP_DEFLATED) as manifest_zip:
+            consumer_export.seek(0)
```
Comment (omaciel) [inline robottelo/manifests.py]: Can you comment on why it is important to "rewind the file" twice (ie. the two `consumer_export.seek(0)` lines? I know what you are doing but not the why :)

File: robottelo/manifests.py [hunk 4]
```diff
@@ -17,160 +12,103 @@
...
+                consumer_export.read()
+            )
+            consumer_export.seek(0)
+            manifest_zip.writestr(
+                'signature',
+                self.signature.sign(SHA256.new(consumer_export.read()))
+            )
+        # Make sure that the file-like object is at the beginning and
+        # ready to be read.
+        manifest.seek(0)
```
Comment (omaciel) [inline robottelo/manifests.py]: Why is this necessary? When you return a file-like object, is it necessary to "rewind" it?

File: robottelo/manifests.py [hunk 5]
```diff
@@ -17,160 +12,103 @@
...
+        return manifest
+
+
+_manifest_cloner = ManifestCloner()
```
Comment (omaciel) [inline robottelo/manifests.py]: Document the reason for doing this?

File: robottelo/ssh.py [hunk 6]
```diff
@@ -95,17 +95,22 @@ def _get_connection(
...
+        ``server.hostname`` from the configuration.
     """
-    if not remote_file:
-        remote_file = local_file
     with _get_connection(hostname=hostname) as connection:
         try:
             sftp = connection.open_sftp()
-            sftp.put(local_file, remote_file)
+            if hasattr(local_file, 'read'):
+                sftp.putfo(local_file, remote_file)
```
Comment (omaciel) [inline robottelo/ssh.py]: Document the need for `sftp.putfo` vs `sftp.put`?

File: robottelo/ui/subscription.py [hunk 7]
```diff
@@ -30,9 +31,12 @@ def upload(self, path, repo_url=None):
             self.text_field_update(locators['subs.repo_url_update'], repo_url)
             self.click(common_locators['save'])
         browse_element = self.wait_until_element(locators['subs.file_path'])
-        browse_element.send_keys(path)
+        with open(manifest.filename, 'wb') as handler:
```
Comment (omaciel) [inline robottelo/ui/subscription.py]: I'm not sure I follow this... are we taking a file-like object and writing it to disk locally (to the Satellite server)? If so, maybe comment here?

File: robottelo/manifests.py (hunk 1 above)
Comment (elyezer) [inline robottelo/manifests.py]: We will not support DSA unless the way manifest content are signed uses it. This signature is being used to just sign the contents of the manifest.

File: robottelo/manifests.py (hunk 2 above)
Comment (elyezer) [inline robottelo/manifests.py]: Will do, but the code already says that since "some_zip.read == extract the file and return its contents". The usage of BytesIO throughout this code is just to avoid file creation.

File: robottelo/manifests.py (hunk 4 above)
Comment (elyezer) [inline robottelo/manifests.py]:
```
In [5]: f = BytesIO()

In [7]: f.write(b'abcd')
Out[7]: 4

In [9]: f.tell()
Out[9]: 4

In [10]: f.seek(0)
Out[10]: 0

In [11]: f.tell()
Out[11]: 0

In [12]: f.seek(4)
Out[12]: 4

In [13]: f.read()
Out[13]: b''

In [14]: f.seek(0)
Out[14]: 0

In [15]: f.read()
Out[15]: b'abcd'
```

File: robottelo/manifests.py (hunk 3 above)
Comment (elyezer) [inline robottelo/manifests.py]: Same as https://github.com/SatelliteQE/robottelo/pull/3201/files#r48831935

File: robottelo/ssh.py (hunk 6 above)
Comment (elyezer) [inline robottelo/ssh.py]: if hasattr(local_file, 'read') == if local_file is a file-like object?
Then putfo (put file object)
Else put (file path)

File: robottelo/ui/subscription.py (hunk 7 above)
Comment (elyezer) [inline robottelo/ui/subscription.py]: Every open operation will happen on the machine where robottelo is running. In order to do a file operation to the Satellite server an ssh operation is needed.

**Comment by elyezer** [comment]
@omaciel thanks for all the comments. Have added comments where you suggested and left a comment with why I don't added a comment at first.

I have answered all other questions, let me know if any other code change is needed.

**Comment by elyezer** [comment]
I'm marking this as do not merge since I'm suspecting that this approach may cause memory leak.

**Comment by elyezer** [comment]
PR is ready.

**Comment by elyezer** [comment]
Re-run the tests

```
$ py.test tests/foreman/api/test_subscription.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 3 items 

tests/foreman/api/test_subscription.py ...

============================= 3 passed in 65.40 seconds =============================

$ py.test tests/foreman/ui/test_subscription.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 3 items 

tests/foreman/ui/test_subscription.py ...

============================ 3 passed in 206.04 seconds =============================

$ py.test tests/foreman/cli/test_subscription.py
================================ test session starts ================================
platform linux -- Python 3.4.2, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/elyezer/code/robottelo, inifile: 
plugins: xdist-1.13.1, cov-2.2.0
collected 6 items 

tests/foreman/cli/test_subscription.py sF....

================== 1 failed, 4 passed, 1 skipped in 165.02 seconds ==================
```

File: robottelo/ui/subscription.py (hunk 7 above)
Comment (omaciel) [inline robottelo/ui/subscription.py]: @elyezer so the answer to my question is "yes"? :)

**Comment by omaciel** [comment]
ACK

**Comment by sthirugn** [comment]
the code looks sane to me.  ACK.  I will also check the next jenkins run to make sure all is good.
